rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,260 @@
+
+Submitting an access request approval does not require an `update` allow rule,
+but since it can be used to escalate privileges, it will still be considered an
+admin action.
+
+Submitting an access request denial will remain a non-admin action.
+
+#### User Actions
+
+Some actions which create, update, or delete user specific resources are
+authorized based on the identity of the requesting user rather than their
+allow rules. For example:
+
+- Creating, listing, and deleting a user's own access requests.
+- Adding or removing MFA devices.
+- Changing the user's password.
+
+These actions will continue to be non-administrative actions and will not require
+MFA to complete.
+
+### UX
+
+The UX of this feature will be very similar to Per-session MFA for `tsh`, `tctl`,
+the WebUI, and Teleport Connect.
+
+When a user performs an admin action, they will be prompted to tap their MFA key.
+Each admin action should require a single tap.
+
+```console
+$ tctl edit roles/access
+# edit role in vim
+Tap any security key
+# success
+```
+
+### MFA enforcement
+
+MFA verification will be required to carry out various administrative actions.
+This will be enforced by the Auth Server API for specific endpoints which are
+considered to be [administrative actions](#administrative-actions).
+
+Essentially, this means that an administrative request to the Auth server will
+follow this flow:
+
+1. Create an Auth API client using the user's existing valid certificates.
+1. Use the client to retrieve an MFA challenge for the user with
+   `rpc CreateAuthenticateChallenge`.
+1. Prompt the user to solve the MFA challenge with their MFA device.
+1. Send the resulting `MFAChallengeResponse` to the Auth Server as part of an
+   administrative API request.
+1. Validate and consume the `MFAChallengeResponse` from the Auth Server to
+   authorize the request (in additional to normal identity-based authorization).
+
+Steps 1-3 are already possible with Teleport currently and used for various MFA
+features. Steps 4 and 5 will require some changes to the Auth API client and
+server respectively.
+
+#### Server changes
+
+For admin actions, the Auth server will validate MFA for the request, either
+from an `MFAChallengeResponse` passed by the client or an MFA-verified certificate
+used for the request. If the request fails MFA verification, an access denied
+error will be returned to the client.
+
+```go
+// ErrAPIMFARequired is returned by AccessChecker when an API request
+// requires an MFA check.
+var ErrAPIMFARequired = trace.AccessDenied("API request requires MFA")
+```
+
+#### Client changes
+
+If the client makes an API request that is considered an admin action, it will
+received the access denied error above. At this point, the client should check
+the error and retry the request with MFA verification.
+
+There are a few different ways that the Auth client can pass MFA verification with
+the Auth server, each with their own pros and cons:
```

Comment:
Reviewers, please help me choose from these options if you have any opinion on the matter. @codingllama

Reply:
I'd favor 1) or 2), as in pass the challenge to all requests either implicitly or explicitly. Part of me likes the "dumb" add-to-all-messages approach, as it is very easy to reason about, but I can see how the metadata-based transport could be made to be an elegant solution.

I have the same concerns about encoding the info into the certs - the certs may live too long, meaning that even if we encoded the purpose of the action, as discussed in various other threads, one could still do many instances of the _same_ action unless we do a bunch of additional bookkeeping. The one-shot challenges won't have this problem, so it's easier to reason about abuse (or lack thereof).

Reply:
@codingllama How strongly do you feel about 1/2 vs 3? My main reason with slightly preferring "3" is that we won't be inventing new authentication protocols and use the same certificate-based approach we've used for years. Similar reason to why we wanted to standardize on the singular approach for propagating client IPs when working on IP pinning instead of having each path do it in its own way.

Reply:
We can pull off any of the 3, but it looks to me that options 1) or 2) are a better fit what we are trying to achieve.

"One-shot" authn is exactly what we want here: no need to generate and manage certs, no fears of exfiltration, no fears of undue reuse, no need to keep tabs on the user doing extra (or multiple of the same!) actions with the same cert, so on. I do think that departing from cert-based isn't necessarily a downside, especially if Auth is in the critical path already.

I won't oppose 3) if y'all rather go for it, but I can't say it looks like the better choice to me.

Reply:
Okay fair enough. In that case I would vote for 1 to err on the side of being more explicit, passing things via client metadata sounds like an easy thing to miss and a bit too implicit for a security oriented feature. This is getting into implementation realm, but I would consider making a separate auth interface that would support MFA auth and wrap the existing one instead of adding an extra parameter to every single method, then the intent will also be clear (similar to ServerWithRoles).

@Joerger @codingllama @rosstimothy WDYGT?

Reply:
Sure, alright by me. I don't totally follow the impl idea, but I trust Brian to follow through.

Reply:
I am happy to go with Option 1. I'll see what I can do about handling MFA Auth in an uninvasive way.

Ideally, we would just need to add a new common `MFAChallengeResponseForAdminAction` rpc message to flip an API request into an admin action. No internal changes in `ServerWithRoles` required.